### PR TITLE
version bump

### PR DESCRIPTION
**Pull request by johannahaffner**
As discussed in https://github.com/patrick-kidger/optimistix/pull/119

### CODE REVIEW

**Comment by johannahaffner** [comment]
Hang on, let me bump up the Python versions to 3.10/3.12 as well

**Comment by patrick-kidger** [comment]
Hmm, a segfault isn't very encouraging.

Looks like it's coming from XLA. Something to do with this version of jaxlib?

**Comment by johannahaffner** [comment]
> Hmm, a segfault isn't very encouraging. 
> Looks like it's coming from XLA. Something to do with this version of jaxlib?

indeed it is not. I had verified that all tests ran locally after bumping the minimum versions of jax, equinox and jaxtyping, so I have a working setup to compare to. I will do that tomorrow.

**Comment by johannahaffner** [comment]
I cannot reproduce this locally with Python 3.12.9. Since the error is the same, I have not tried yet with 3.10. 

All configurations should be the same, see for local (first) and Github tests (second)

``` 
> pip freeze
beartype==0.20.0
cfgv==3.4.0
distlib==0.3.9
equinox==0.11.12
execnet==2.1.1
filelock==3.17.0
identify==2.6.8
iniconfig==2.0.0
jax==0.5.2
jaxlib==0.5.1
jaxtyping==0.2.38
-e git+https://github.com/johannahaffner/lineax.git@ccdbdbd8e88fb014d22d4fcfce362a010b4c6d97#egg=lineax
ml_dtypes==0.5.1
nodeenv==1.9.1
numpy==2.2.3
opt_einsum==3.4.0
packaging==24.2
platformdirs==4.3.6
pluggy==1.5.0
pre_commit==4.1.0
pytest==8.3.5
pytest-xdist==3.6.1
PyYAML==6.0.2
scipy==1.15.2
typeguard==2.13.3
typing_extensions==4.12.2
uv==0.6.5
virtualenv==20.29.3
wadler_lindig==0.1.3
```

For comparison, this is what we get in here: 

```
Run python -m pip install .
Processing /home/runner/work/lineax/lineax
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Requirement already satisfied: equinox>=0.11.10 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from lineax==0.0.8) (0.11.12)
Requirement already satisfied: jax>=0.4.38 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from lineax==0.0.8) (0.5.2)
Requirement already satisfied: jaxtyping>=0.2.24 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from lineax==0.0.8) (0.2.38)
Requirement already satisfied: typing-extensions>=4.5.0 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from lineax==0.0.8) (4.12.2)
Requirement already satisfied: wadler-lindig>=0.1.0 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from equinox>=0.11.10->lineax==0.0.8) (0.1.3)
Requirement already satisfied: jaxlib<=0.5.2,>=0.5.1 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from jax>=0.4.38->lineax==0.0.8) (0.5.1)
Requirement already satisfied: ml_dtypes>=0.4.0 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from jax>=0.4.38->lineax==0.0.8) (0.5.1)
Requirement already satisfied: numpy>=1.25 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from jax>=0.4.38->lineax==0.0.8) (2.2.3)
Requirement already satisfied: opt_einsum in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from jax>=0.4.38->lineax==0.0.8) (3.4.0)
Requirement already satisfied: scipy>=1.11.1 in /opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages 
(from jax>=0.4.38->lineax==0.0.8) (1.15.2)
Building wheels for collected packages: lineax
  Building wheel for lineax (pyproject.toml): started
  Building wheel for lineax (pyproject.toml): finished with status 'done'
  Created wheel for lineax: filename=lineax-0.0.8-py3-none-any.whl size=67341 sha256=9212ec86ade2561829e589d0bf3fba2c98dfd0d8cbcf179101886ae44d9a7975
  Stored in directory: /tmp/pip-ephem-wheel-cache-6f71lm04/wheels/13/ad/35/16376dd929f1b621f16819f78851ed90ec8e38cd8625496139
Successfully built lineax
Installing collected packages: lineax
Successfully installed lineax-0.0.8
```

**Comment by patrick-kidger** [comment]
So we're running on Linux in CI. If you're on something else locally (MacOS?) then that might matter for an error of this type. Can you try on a Linux server somewhere?

In addition, it looks like it is specifically `tests/test_jvp_jvp1.py` that fails. Try deleting all the other tests and running just that one through CI with the pytest `-vvv` flag and if we're lucky this might be possible to narrow down to a single test (I'm hoping it's not something due to an interaction between multiple tests :D ) That'll likely give us a good signal to identify what might be wrong.

**Comment by johannahaffner** [comment]
I can run this on a linux server, yes. Will do that at some point tomorrow!

**Comment by johannahaffner** [comment]
Did not mean to close this. Looks like I can't stop the jobs from running and sending you emails when they fail. Sorry about this!

**Comment by johannahaffner** [comment]
So I did manage to reduce this! Something goes amiss when using 

https://github.com/patrick-kidger/lineax/blob/52f9d3f78e153125e76e35f508d9ad97e3013e70/tests/helpers.py#L31

with JAX 0.5.1, and the corresponding jaxlib version. Everything is well behaved up until 0.5.0. `_construct_matrix_impl` has an `ft.lru_cache` decorator, and indeed there seems to be going wrong when caching.

Two things that don't help: 
- changing the maximum size of the cache. The error persists even with `max_size=2`. 
- bisecting through the list of solvers. It seems that it is the number of solvers, rather than any specific one, that causes the error.

Full error message for completeness:

```
============================= test session starts ==============================
platform linux -- Python 3.12.8, pytest-8.3.5, pluggy-1.5.0
rootdir: /home/jhaffner/lineax
configfile: pyproject.toml
plugins: jaxtyping-0.2.38, xdist-3.6.1
collected 20 items                                                             

tests/test_jvp_jvp1.py ...................Fatal Python error: Segmentation fault

Thread 0x00007f2614518740 (most recent call first):
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/compiler.py", line 324 in backend_compile
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/profiler.py", line 334 in wrapper
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/compiler.py", line 690 in _compile_and_write_cache
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/compiler.py", line 473 in compile_or_get_cached
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/interpreters/pxla.py", line 2757 in _cached_compilation
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/interpreters/pxla.py", line 2959 in from_hlo
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/interpreters/pxla.py", line 2446 in compile
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/pjit.py", line 1657 in _pjit_call_impl_python
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/pjit.py", line 195 in _python_pjit_helper
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/pjit.py", line 341 in cache_miss
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/jax/_src/traceback_util.py", line 180 in reraise_with_filtered_traceback
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/equinox/_jit.py", line 244 in _call
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/equinox/_module.py", line 1096 in __call__
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/equinox/_jit.py", line 275 in __call__
  File "/home/jhaffner/lineax/tests/helpers.py", line 449 in jvp_jvp_impl
  File "/home/jhaffner/lineax/tests/test_jvp_jvp1.py", line 36 in test_jvp_jvp
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/python.py", line 159 in pytest_pyfunc_call
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/python.py", line 1627 in runtest
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/runner.py", line 174 in pytest_runtest_call
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/runner.py", line 242 in <lambda>
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/runner.py", line 341 in from_call
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/runner.py", line 241 in call_and_report
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/runner.py", line 132 in runtestprotocol
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/runner.py", line 113 in pytest_runtest_protocol
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/main.py", line 362 in pytest_runtestloop
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/main.py", line 337 in _main
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/config/__init__.py", line 175 in main
  File "/home/jhaffner/.lineax-test/lib/python3.12/site-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/home/jhaffner/.lineax-test/bin/pytest", line 10 in <module>

Extension modules: numpy._core._multiarray_umath, numpy.linalg._umath_linalg, jaxlib.cpu_feature_guard, scipy._lib._ccallback_c, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg.cython_lapack, scipy.linalg._cythonized_array_utils, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, scipy.linalg._solve_toeplitz, scipy.linalg._decomp_lu_cython, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg._matfuncs_expm, scipy.linalg._linalg_pythran, scipy.linalg.cython_blas, scipy.linalg._decomp_update, scipy.sparse._sparsetools, _csparsetools, scipy.sparse._csparsetools, scipy.sparse.linalg._dsolve._superlu, scipy.sparse.linalg._eigen.arpack._arpack, scipy.sparse.linalg._propack._spropack, scipy.sparse.linalg._propack._dpropack, scipy.sparse.linalg._propack._cpropack, scipy.sparse.linalg._propack._zpropack, scipy.sparse.csgraph._tools, scipy.sparse.csgraph._shortest_path, scipy.sparse.csgraph._traversal, scipy.sparse.csgraph._min_spanning_tree, scipy.sparse.csgraph._flow, scipy.sparse.csgraph._matching, scipy.sparse.csgraph._reordering (total: 40)
Segmentation fault (core dumped)
```

**Comment by patrick-kidger** [comment]
Aha awesome! Progress. Add on a `-vvv` and it should tell you which solver it crashes on.

Odd the bisecting the list of solvers doesn't help. This could be because of a few weird possibilities -- a particular solver being used after another particular solver; at least N solvers being used, etc. You could try permuting the list of solvers and seeing if that yields any variability on where it crashes.

**Comment by johannahaffner** [comment]
Gosh that is useful! Should have tried that straight away.

It always fails for the 19th solver. I shuffled the list with random.shuffle, tried three combinations - each time a different solver is in 19th place. It fails for Cholesky, AutoLinearSolver, and BICGStab.

**Comment by johannahaffner** [comment]
AutoLinearSolver without tags selects LU/SVD depending on whether the operator is well-posed. I can go back to check which one it is later, but I think three establishes a trend.

**Comment by johannahaffner** [comment]
Clearing caches before running the tests allows everything to run again on Linux. 

I'm not familiar enough with the details of lineax autodiff and custom primitives (for the transposition rules?) - do you think it could be due to these? If not, then I can open a JAX issue and add my bisected test file (20/160 tests) as an MWE of sorts.

**Comment by patrick-kidger** [comment]
Hmm, do we still need Lineax to make this trigger? And if so, which bits of Lineax? The fact that it's consistently at 19 suggests that it'll be triggered by something that is consistent across all solvers, like 'how many pure callbacks are registered' / 'how many jit'd functions are compiled'.

My guess is that the actual body of the test could be dummied out with something else.

**Comment by johannahaffner** [comment]
I think we do, when I replace it with `jnp.linalg.solve`, then we no longer get the segfault. I'll keep digging.

**Comment by patrick-kidger** [comment]
Btw, when you tweak things appropriately here -- can you target `dev` instead of `main`? We've still got a few unreleased things on that branch.

File: pyproject.toml
Comment (patrick-kidger) [inline pyproject.toml:1]: Heads-up that I've just pushed to `dev` with modifications to `pyproject.toml`; you might see some (easy to resolve) merge conflicts.

Also if we bump the dependencies these should also be updated in `README.md` and `docs/index.md`. (Although I've not been very good at updating these myself previously...)

**Comment by johannahaffner** [comment]
Now beartype is not happy, there have not been any changes to the function in question. Probably the new version is a little more stringent for this specific case. I'll check it out.

**Comment by patrick-kidger** [comment]
Okay awesome! It looks like this is all passing now :D 

In that case time to merge this I think, and then move on to doing the `dev->main` release!

Thanks a lot for all your efforts pushing this through Johanna. What was supposed to be a simple version bump turned into a very complicated debugging experience, and I really appreciate you putting in the time to figure this out.

**Comment by johannahaffner** [comment]
Thank you! I'm glad to see it out in the wider world 🙃
